turbopack/crates/turbo-tasks-malloc/src/counter.rs

```diff
@@ -1,124 +1,198 @@
+                    self.counters.deallocation_count.load(Ordering::Relaxed);
+                guard.allocations += self.counters.allocations.load(Ordering::Relaxed);
+                guard.deallocations += self.counters.deallocations.load(Ordering::Relaxed);
+            }
+            self.registered = false;
+            self.counters = AtomicCounters::new();
         }
-        self.allocation_counters = AllocationCounters::default();
+    }
+}
+
+/// For the most parts threads managed by tokio have [unload] specifically called, so this `drop`
+/// impl is really for threads not managed by tokio (rayon, chili, etc). Without we would leak
+/// memory in those cases.
+impl Drop for ThreadLocalCounter {
+    fn drop(&mut self) {
+        self.unload()
     }
 }
 
 thread_local! {
   static LOCAL_COUNTER: UnsafeCell<ThreadLocalCounter> = const {UnsafeCell::new(ThreadLocalCounter::new())};
 }
 
-pub fn get() -> usize {
-    ALLOCATED.load(Ordering::Relaxed)
+// stores an estimate of the peak memory allocated.
+static MAX_ALLOCATED: AtomicUsize = AtomicUsize::new(0);
+
+/// Returns an estimate of the total memory statistics
+pub fn global_counters() -> AllocationCounters {
+    let mut counters = AllocationCounters::new();
+    {
+        let guard = ACTIVE.lock().unwrap();
+        counters.allocation_count = guard.allocation_count;
+        counters.deallocation_count = guard.deallocation_count;
+        counters.allocations = guard.allocations;
+        counters.deallocations = guard.deallocations;
+        for &thread in guard.active.iter() {
+            // SAFETY: the ThreadLocalAllocationCounters cannot be dropped without grabbing the
+            // lock and we are holding the lock right now so this reference is guaranteed to be
+            // valid.
+            let thread = unsafe { &*thread.0 };
+            // TODO re-evaluate memory ordering.
+            counters.allocation_count += thread.allocation_count.load(Ordering::Acquire);
+            counters.deallocation_count += thread.deallocation_count.load(Ordering::Acquire);
+            counters.allocations += thread.allocations.load(Ordering::Acquire);
+            counters.deallocations += thread.deallocations.load(Ordering::Acquire);
+        }
+    }
+
+    MAX_ALLOCATED.fetch_max(
+        counters.allocations - counters.deallocations,
+        Ordering::AcqRel,
+    );
+    counters
 }
 
 pub fn allocation_counters() -> AllocationCounters {
-    with_local_counter(|local| local.allocation_counters.clone())
+    with_local_counter(|local| local.counters.to_allocation_counters())
 }
 
+/// Resets the counters for the current thread.
+/// This is used to exclude some work from the metrics and as such should be used sparingly.
 pub fn reset_allocation_counters(start: AllocationCounters) {
-    with_local_counter(|local| local.allocation_counters = start);
+    with_local_counter(|local| local.counters = AtomicCounters::from_allocation_counters(start));
```

Comment:
I feel like this can break something as it doesn't use atomic operations to modify the counters.